### ISSUE TITLE
Customer facing prices for Mix and Match Products

### ISSUE THREAD

**Issue by helgatheviking** [Automattic/woocommerce-payments#2974]
### Description

The [Mix and Match](https://woocommerce.com/products/woocommerce-mix-and-match-products/) plugin enables a customer to select X number of products to fill a container.... think Grouped products with a quantity constraint. Optionally the product price can be dynamic and dependent on the prices of the products selected. Ideally, I'd like to see this price entered be the price that the payment request buttons use.. and enable/disable the request buttons when the container configuration is not valid... and send all the chosen products to the processor as well. 

In the short-term, I could get by with a way to disable the display of the request buttons on products that do not support them.

### Acceptance criteria

Long term
- [ ] Enable/disable buttons when price is/not valid
- [ ] Send dynamic price to payment processor
- [ ] Send selected products to the payment processor

Short term
- [x] PHP filter to disable display of buttons where not supported [added in 3.4](https://github.com/Automattic/woocommerce-payments/blob/e6df2868297327ba42c265da8c7b374bab0e11e2/changelog.txt#L1847)

### Testing instructions

1. Create a Mix and Match product
2. Click on request buttons (apple/google pay)
3. presumably see $0 price regardless of what price you've entered. 
4. presumably do not see the selected options*

*it's getting late so I haven't explicitly tested this yet, but was on a roll from reporting the Name Your Price issue in #2973. The issue of dynamic pricing is pretty similar and related to some other compatibility issues you already have.

**Comment by vbelolapotkov** [comment]
@pierorocca flagging this one with you since this one is related to payment request buttons and compatibility with a plugin in Woo marketplace.

**Comment by pierorocca** [comment]
@FangedParakeet relevant to the tokenized carts effort, the ECE migration or needing its own solve?

**Comment by FangedParakeet** [comment]
This is pretty relevant to the tokenized carts effort. I'm not sure that this would immediately be solved by tokenized carts alone, but in our initial implementation in #8526, I believe we are planning to include pathways for third-party plugins to adjust final cart amounts prior to proceeding with the express checkout flow.

> In the short-term, I could get by with a way to disable the display of the request buttons on products that do not support them.

This is an interesting idea. @pierorocca, do you think this functionality would be something we would be interested in opening up to third-party plugins? That functionality being giving third-party plugins the ability and empowerment to disable our express checkout methods on products where they deem they will not be compatible. Alternative options could be to declare incompatibility with these extensions in our own plugin settings or simply to not act at all.

**Comment by helgatheviking** [comment]
FWIW, conditionally disabling buttons was [added in 3.4](https://github.com/Automattic/woocommerce-payments/blob/e6df2868297327ba42c265da8c7b374bab0e11e2/changelog.txt#L1847). 

Just to clarify, "Tokenized carts" means you'd generate a standalone cart and send that cart to the request service?

**Comment by pierorocca** [comment]
> That functionality being giving third-party plugins the ability and empowerment to disable our express checkout methods on products where they deem they will not be compatible. Alternative options could be to declare incompatibility with these extensions in our own plugin settings or simply to not act at all.

I'm not a fan of having the 3P extension make the final call if the buttons should be hidden/disabled. I like the idea of having extensions declare compatibility/incompatibility and having WooCommerce orchestrate what should be done with that info like keep enabled and notify merchant of an incompatibility, disable and notify merchant, etc. This way we can also have the compatibility documented and known before installation and made fully transparent. If a merchant installs an extension and all of a sudden things starts disappearing, it'll be very hard for a merchant to understand what's going on.

**Comment by pierorocca** [comment]
> * Add - Introduce `wcpay_payment_request_is_product_supported` filter. Allow plugins to conditionally disable payment request buttons on products that do not support them.

Yes, not a supporter of this approach, implemented before my time, for reasons stated above. It's also an approach that opens up the functionality to abuse should a bad actor want to suppress competitive options.

**Comment by pierorocca** [comment]
One last comment, WooPay unlike other express checkout options, we've done a ton of work to achieve compatibility that aren't possible with other express checkout options. WooPay is compatible with Mix and Match. 

@FangedParakeet does the filter disable all PRBs including WooPay? If so the approach is flawed in scenarios where some express checkout options are compatible and where some are not.

**Comment by helgatheviking** [comment]
> I like the idea of having extensions declare compatibility/incompatibility and having WooCommerce orchestrate what should be done with that info like keep enabled and notify merchant of an incompatibility, disable and notify merchant, etc. 

I mean that sounds nice, but we needed _something_ that was going to prevent my users' customers from launching the PBR flow when it didn't work with my product type. Customers were checking out for $0 and no items were selected. It was a bad experience. I'm certainly open to improvements.

>WooPay is compatible with Mix and Match.

Is it? I'll admit I have not circled back to this compatibility since I disabled the buttons on a mix and match single product page.. which coincides with your 3.4 release.

I was under the impression dynamic prices (and adding more than one prodct to the cart) required tokenized cart being complete? But again, I'm only following development through this and any other issues I've opened. 

If it's quite a bit more robust now in what the single product page can handle I would want to retest and remove my button disabling filter.

**Comment by pierorocca** [comment]
> I mean that sounds nice, but we needed something 

Of course :) That was 3 years ago and who wants to wait that long?

Regarding [WooPay](https://woocommerce.com/woopay/), which is a newish Express Checkout feature of WooPayments, and that's offered alongside Google Pay and Apple Pay, our testing shows it's compatible and would love to get your confirmation @helgatheviking. 

![image](https://github.com/Automattic/woocommerce-payments/assets/109928953/3707d1e4-94c3-4312-9870-017560ad1156)

![image](https://github.com/Automattic/woocommerce-payments/assets/109928953/3ad9268c-3333-45b9-a142-0340121513d8)

![image](https://github.com/Automattic/woocommerce-payments/assets/109928953/2570e720-fa6f-4213-8b81-f9107a1d03db)

![image](https://github.com/Automattic/woocommerce-payments/assets/109928953/e041da4f-67f1-487c-8b45-69839dcabb6d)


One thing I see as an enhancement opportunity is reducing the "Selections :" from 1 per selection to just 1 occurrence in the WooPay order summary.

cc @alefesouza.

**Comment by pierorocca** [comment]
https://proccaatomic.wpcomstaging.com/ is my test store should you like to try it out.

**Comment by pierorocca** [comment]
Screen capture resulting from clicking "Buy with WooPay" on the PDP page:

![image](https://github.com/Automattic/woocommerce-payments/assets/109928953/4a32df96-b696-48ab-b867-3db62413b0e1)

**Comment by helgatheviking** [comment]
@pierorocca that's incredible!  i like to think I am _extremely_ informed about all things woo, but I didn't even know about this. how cool!  So now I would definitely want to make a distinction between the button types that I disable.

I am going to guess if it supports MNM is will support my other plugin, Name Your Price? Arguably that's simpler.

**Comment by pierorocca** [comment]
It is compatible :) https://woocommerce.com/document/woopay-merchant-documentation/#compatible-extensions

**Comment by FangedParakeet** [comment]
> does the filter disable all PRBs including WooPay? If so the approach is flawed in scenarios where some express checkout options are compatible and where some are not.

AFAICT, the filter will only disable our PRB buttons (i.e. Apple/Google Pay), but will not affect WooPay. 👍

**Comment by pierorocca** [comment]
@helgatheviking I'd like to learn what channels you use to keep up with all things Woo. I'm likely out of the loop and WooPay is one thing I'd like to get the word out more than I have. Mind emailing me at piero.rocca@automattic.com?

**Comment by helgatheviking** [comment]
@pierorocca I did email you, but not sure if you received it. Generally the community slack, marketplace vendor slack, dev blog, and Twitter.
